daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExceptionTest.scala

```diff
@@ -0,0 +1,536 @@
+import com.daml.lf.testing.parser.Implicits._
+import com.daml.lf.validation.Validation
+import org.scalatest.prop.TableDrivenPropertyChecks
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class ExceptionTest extends AnyWordSpec with Matchers with TableDrivenPropertyChecks {
+
+  "builtin exception types: constructors & destructors" should {
+
+    val pkgs: PureCompiledPackages = typeAndCompile(p"""
+       module M {
+         val text : Text = "Hello!";
+         val t2 : Text = GENERAL_ERROR_MESSAGE (MAKE_GENERAL_ERROR M:text);
+         val t3 : Text = ARITHMETIC_ERROR_MESSAGE (MAKE_ARITHMETIC_ERROR M:text);
+         val t4 : Text = CONTRACT_ERROR_MESSAGE (MAKE_CONTRACT_ERROR M:text);
+       }
+      """)
+
+    val testCases = Table[String, String](
+      ("expression", "string-result"),
+      ("M:text", "Hello!"),
+      ("M:t2", "Hello!"),
+      ("M:t3", "Hello!"),
+      ("M:t4", "Hello!"),
+    )
+
+    forEvery(testCases) { (exp: String, res: String) =>
+      s"""eval[$exp] --> "$res"""" in {
+        val expected: SResult = SResultFinalValue(SValue.SText(res))
+        runExpr(pkgs)(e"$exp") shouldBe expected
+      }
+    }
+  }
+
+  "builtin exceptions: to/from-any-exception" should {
+
+    val pkgs: PureCompiledPackages = typeAndCompile(p"""
+       module M {
+         val G : AnyException = to_any_exception @GeneralError (MAKE_GENERAL_ERROR "G");
+         val A : AnyException = to_any_exception @ArithmeticError (MAKE_ARITHMETIC_ERROR "A");
+         val C : AnyException = to_any_exception @ContractError (MAKE_CONTRACT_ERROR "C");
+         val fromG : AnyException -> Text = \(e:AnyException) -> case from_any_exception @GeneralError e of None -> "NONE" | Some x -> GENERAL_ERROR_MESSAGE x;
+         val fromA : AnyException -> Text = \(e:AnyException) -> case from_any_exception @ArithmeticError e of None -> "NONE" | Some x -> ARITHMETIC_ERROR_MESSAGE x;
+         val fromC : AnyException -> Text = \(e:AnyException) -> case from_any_exception @ContractError e of None -> "NONE" | Some x -> CONTRACT_ERROR_MESSAGE x;
+       }
+      """)
+
+    val testCases = Table[String, String](
+      ("expression", "string-result"),
+      ("M:fromG M:G", "G"),
+      ("M:fromA M:A", "A"),
+      ("M:fromC M:C", "C"),
+      ("M:fromG M:A", "NONE"),
+      ("M:fromG M:C", "NONE"),
+      ("M:fromA M:G", "NONE"),
+      ("M:fromA M:C", "NONE"),
+      ("M:fromC M:G", "NONE"),
+      ("M:fromC M:A", "NONE"),
+    )
+
+    forEvery(testCases) { (exp: String, res: String) =>
+      s"""eval[$exp] --> "$res"""" in {
+        val expected: SResult = SResultFinalValue(SValue.SText(res))
+        runExpr(pkgs)(e"$exp") shouldBe expected
+      }
+    }
+  }
+
+  "ANY_EXCEPTION_MESSAGE" should {
+
+    val pkgs: PureCompiledPackages = typeAndCompile(p"""
+       module M {
+
+         record @serializable MyException1 = { message: Text } ;
+         exception MyException1 = {
```

Comment:
Following my comment above, it would be nice to have test cases where `message` crashes. My current understanding is that in that case `to_any_exception` will not crash so ignoring the potential issue around unhandled exceptions that isn’t going to change the exception that gets thrown.

Reply:
we do test this!